[MirAL] Implement a "no active window" state

peat-psuwit opened the pull request:
When setting the active window to a null window, mru_active_windows
won't be updated. This commit add the `allow_active_window` flag which
will be false when a null window is set, disabling the active window.

This is used by Lomiri during spreads and other effects and animations.

Co-authored-by: Alan Griffiths <alan@octopull.co.uk>

This PR comes with Jenkinsfile update.

Fixes ubports/unity8#225.

peat-psuwit committed:
Update Jenkinsfile as of ubports/build-tools@10c0987

peat-psuwit commented:
Well, it crashes unity8 when a trusted surface is closed. Will investigate further.

Flohack74 commented:
On an unrelated note, I still have situations in TELEports when it thinks that the App is still suspended... It got better with the other commit, but is not entirely gone.

peat-psuwit committed:
[MirAL] Implement a "no active window" state

When setting the active window to a null window, mru_active_windows
won't be updated. This commit add the `allow_active_window` flag which
will be false when a null window is set, disabling the active window.

This is used by Lomiri during spreads and other effects and animations.

Co-authored-by: Alan Griffiths <alan@octopull.co.uk>

peat-psuwit committed:
[MirAL] always use active_window() to find out active window

BasicWindowManager::remove_window() was using mru_active_windows to find
out if the removing window is the active window. To be consistent with
the rest of the code, it's changed to use active_window().

peat-psuwit commented:
The crash is fixed. Now ready for review.

Flohack74 commented on src/miral/basic_window_manager.cpp:
```diff
@@ -511,7 +511,17 @@ void miral::BasicWindowManager::force_close(Window const& window)
 auto miral::BasicWindowManager::active_window() const -> Window
 {
-    return mru_active_windows.top();
+    auto const focussed_surface = focus_controller->focused_surface();
```
Small typo, it would be focused_surface in English :)

peat-psuwit replied:
Well, this code is copied from another code in the same file. Also, "focussed" is kind of a valid spelling in British English (and Canonical is a UK-based company so... :) ).

https://en.wiktionary.org/wiki/focussed

Flohack74 replied:
Its just odd that in the same line of code you have 2 spellings... If you want to keep it though I wont block it ofc.

peat-psuwit replied:
Oh, didn't notice that there're 2 spellings in the same line. Then I'll fix it.

mariogrip commented:
Has this been proposed upstream?

Ping @AlanGriffiths

AlanGriffiths commented:
> Has this been proposed upstream?
> 
> Ping @AlanGriffiths

Quick answer:

1. It hasn't been proposed. 
2. It doesn't look stupid (apart from mixing Jenkins stuff with code), but I've not got my head sufficiently around the issue to know if it's right.

peat-psuwit commented:
This PR gets updated with the suggested fix from Alan which seems to be cleaner.

peat-psuwit commented:
Ok. Confirmed that this actually fixes the issue. Ready to be merged!
